soc/nordic/nrf54h/power.c

```diff
@@ -126,20 +125,37 @@ static void do_suspend_to_ram(void)
 
 	sys_resume();
 }
-#endif /* IS_ENABLED(CONFIG_PM_S2RAM) */
+#endif /* defined(CONFIG_PM_S2RAM) */
 
 void pm_state_set(enum pm_state state, uint8_t substate_id)
 {
 	if (state != PM_STATE_SUSPEND_TO_RAM) {
 		k_cpu_idle();
 		return;
 	}
-#if IS_ENABLED(CONFIG_PM_S2RAM)
+#if defined(CONFIG_PM_S2RAM)
 	do_suspend_to_ram();
 #endif
 }
 
 void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 {
+#if !defined(CONFIG_SOC_NRF54H20_CPURAD)
+	if (state == PM_STATE_SUSPEND_TO_IDLE) {
+		soc_lrcconf_poweron_release(&pm_node, NRF_LRCCONF_POWER_DOMAIN_0);
```

Comment:
This only releases the poweron request when `PM_STATE_SUSPEND_TO_IDLE` was used to call `k_cpu_idle()` via the pm subsystem, but if the pm subsystem decided against suspending the system it will call `k_cpu_idle()` [here](https://github.com/nrfconnect/sdk-zephyr/blob/main/kernel/idle.c#L85) instead and then not call the release.
This can happen if the is no `suspend-to-idle` state defined in devicetree or it has a minimum residency time that isn't met, e.g. a `k_msleep(10)` for with this pm configuration https://github.com/nrfconnect/sdk-nrf/blob/84492d6d9264745050a018d01f4f333b285864c1/tests/benchmarks/multicore/idle/boards/nrf54h20dk_nrf54h20_cpuapp_s2ram.overlay#L12C39-L12C39

This is still way better than losing power, but needs to be followed up to have all paths covered eventually

Reply:
there is this mechanism but it is quite ugly and requires special header with a macro. It was added as a workaround for PAN where code need to be executed immediately after WFI (so function cannot be used). I have https://github.com/zephyrproject-rtos/zephyr/pull/71667 which adds function hook symmetric to on_enter hook but PR stuck.

Reply:
Maybe your PR would have a clean usecase now and it could be argued for again to avoid that assembly mess?